File: codeEg/7. 최단 경로/화성 탐사.py
import heapq

def dijkstra():
  q = []
  heapq.heappush(q, (table[0][0], 0, 0))
  
  dx = [1, 0, -1, 0]
  dy = [0, 1, 0, -1]
  while q:
    dist, x, y = heapq.heappop(q)

    if dist > distance[x][y]:
      continue

    for i in range(4):
      nx, ny = x + dx[i], y + dy[i]

      if 0<=nx<n and 0<=ny<n:
        if distance[nx][ny] > distance[x][y] + table[nx][ny]:
          distance[nx][ny] = distance[x][y] + table[nx][ny]
          heapq.heappush(q, (distance[nx][ny], nx, ny))

T = int(input())
for _ in range(T):
  n = int(input())
  table = [list(map(int, input().split())) for _ in range(n)]
  INF = int(1e9)
  distance = [[INF] * n for _ in range(n)]
  distance[0][0] = table[0][0]

  dijkstra()

  print(distance[n-1][n-1])

File: codeEg/7. 최단 경로/화성 탐사.js
"use strict"

function solution(table) {
  const n = table.length

  const INF = 1e9
  const distance = new Array(n).fill().map(() => new Array(n).fill(INF))

  distance[0][0] = table[0][0]
  const q = []
  q.push([table[0][0], 0, 0])

  const dx = [1, 0, -1, 0]
  const dy = [0, 1, 0, -1]
  while (q.length) {
    q.sort((a, b) => a[0] - b[0])
    const [dist, x, y] = q.shift()

    if (dist > distance[x][y]) continue

    for (let i = 0; i < 4; i++) {
      const [nx, ny] = [x + dx[i], y + dy[i]]

      if (0 <= nx && nx < n && 0 <= ny && ny < n) {
        if (distance[nx][ny] > dist + table[nx][ny]) {
          distance[nx][ny] = dist + table[nx][ny]
          q.push([distance[nx][ny], nx, ny])
        }
      }
    }
  }
  return distance[n-1][n-1]
}

const input_data = `
9 0 5 1 1 5 3
4 1 2 1 6 5 3
0 7 6 1 6 8 5
1 1 7 8 3 2 3
9 4 0 7 6 4 1
5 8 3 2 4 8 3
7 4 8 4 8 3 4
`.trim().split('\n')

console.log(solution(input_data.map(el => el.split(' ').map(Number))))